File: API.NEWS/Controllers/AuthController.cs
namespace API.NEWS.Controllers
{
    [ApiController]
    [Route("api/[controller]")]
    public class AuthController : ControllerBase
    {
        private readonly ApplicationDbContext _dbContext;
        private readonly IEmailService _emailservice;
        public AuthController(ApplicationDbContext dbContext, IEmailService emailservice) {
            _dbContext = dbContext;
            _emailservice = emailservice;
        }

        [HttpPost]
        [Route("register")]
        public async Task<IActionResult> Register(UserDto register) {
            //validasi users
            if(_dbContext.Users.Any(u => u.Email == register.Email)) {
                return BadRequest(new {
                status = false,
                message = "Users already exists.",
                data = new {}
                });
            }
            //hash password
            var PasswordHasher = new PasswordHasher<string>();
            var hash = PasswordHasher.HashPassword(register.Email, register.Password);
            //create token
            var randomToken = Convert.ToHexString(RandomNumberGenerator.GetBytes(64));
            //create new data to database
            var NewUser = new User() {
                Email = register.Email,
                Password = hash,
                VerificationToken = randomToken,
            };
            //save to database
            _dbContext.Users.Add(NewUser);
            await _dbContext.SaveChangesAsync();
            //send email 
            var email  = new EmailDto() {
                To = register.Email,
                Subject = "VERIFICATION EMAIL",
                Body = "<br/><br/>We are excited to tell you that your account is" +  
   " successfully created. Please input this code to verify your account" +  
    " <br/><br/>code : " + randomToken + "<br/><br/> Thank you. ",
        };
         _emailservice.SendEmail(email);
            //response
            return Ok(new {
                status = true,
                message = "Register successly, check your email to verification !",
                data = NewUser,
            });
        }

        [HttpPost]
        [Route("login")]
        public async Task<IActionResult> Login(UserDto login) {
            //find user
            var user = _dbContext.Users.FirstOrDefault(x => x.Email == login.Email);
            //validasi users is ready
            if(user == null) {
                return BadRequest(new {
                status = false,
                message = "Users not found.",
                data = new {}
                });
            }
            //verify password hash
            var PasswordHasher = new PasswordHasher<string>();
            var verifPass = PasswordHasher.VerifyHashedPassword(login.Email, user.Password, login.Password);
            if(verifPass == PasswordVerificationResult.Failed) {
                return BadRequest(new {
                status = false,
                message = "Incorrect Password",
                data = new {}
                });
            }
            //validasi users is verified
            if(user.VerifiedAt == null) {
                return BadRequest(new {
                status = false,
                message = "Not verified",
                data = new {}
                });
            }
            // definisikan klaim
            var claims = new List<Claim>
            {
                new Claim("sub", user.Id.ToString()), // sub (subject) adalah claim yang umum digunakan untuk menyatakan ID pengguna
                new Claim("email", user.Email), // contoh klaim email
            };
            // create jwt
            var secureKey = "this is very secure key";
            var symmetric = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(secureKey));
            var credentials = new SigningCredentials(symmetric, SecurityAlgorithms.HmacSha256Signature);
            var header = new JwtHeader(credentials);
            var payload = new JwtPayload (null, null, claims, null, DateTime.Today.AddDays(1));
            var securityToken = new JwtSecurityToken(header, payload);
            var token = new JwtSecurityTokenHandler().WriteToken(securityToken);
            //response
            return Ok(new {
                status = true,
                message = $"Welcome back, {user.Email}",
                data = new {
                    auth = new {
                        token = token,
                        token_type = "Bearer",
                        Expires = DateTime.UtcNow.AddDays(1),
                    },
                    user = user
                }
            });
        }
        
        [HttpPost]
        [Route("verify")]
        public async Task<IActionResult> Verify(String token) {
            //find user
            var user = _dbContext.Users.FirstOrDefault(x => x.VerificationToken == token);
            //validasi user
            if(user == null) {
                return BadRequest(new {
                status = false,
                message = "Users not found.",
                data = new {}
                });
            }
            //update date verified
            user.VerifiedAt = DateTime.UtcNow;
            await _dbContext.SaveChangesAsync();
            //response
            return Ok(new {
                status = true,
                message = "Users verified.",
                data = user,
                });
        }
        
        [HttpPost]
        [Route("forgot-password")]
        public async Task<IActionResult> ForgotPassword(String email) {
            //find user
            var user = _dbContext.Users.FirstOrDefault(x => x.Email== email);
            if(user == null) {
                return BadRequest(new {
                status = false,
                message = "Users not found.",
                data = new {}
                });
            }
            //create random token
             var randomToken = Convert.ToHexString(RandomNumberGenerator.GetBytes(64));
            user.PasswordResetToken = randomToken;
            user.ResetTokenExpires = DateTime.UtcNow.AddDays(1);
            await _dbContext.SaveChangesAsync();
            
            //send email 
             var sendEmail  = new EmailDto() {
                 To = email,
                 Subject = "FORGOT PASSWORD",
                 Body = "<br/><br/>Input the code below to reset your account password." +  
       "  If you didn't request a new password, you can safely delete this email." +  
       " <br/><br/>code : " + randomToken + "<br/><br/> Thank you. ",
             };
             _emailservice.SendEmail(sendEmail);
            return Ok(new {
                status = true,
                message = "you may reset your token now, check your email for get code",
                data = randomToken,
                });
        }
        
        [HttpPost]
        [Route("reset-password")]
        public async Task<IActionResult> ResetPassword(ResetPasswordDto reset) {
            //find user
            var user = _dbContext.Users.FirstOrDefault(x => x.PasswordResetToken == reset.Token);
            if(user == null || user.ResetTokenExpires < DateTime.UtcNow) {
                return BadRequest(new {
                status = false,
                message = "invalid token",
                data = new {}
                });
            }
            //hash password
            var PasswordHasher = new PasswordHasher<string>();
            var hash = PasswordHasher.HashPassword(user.Email, reset.VerifyPassword);
            
            user.Password = hash;
            user.PasswordResetToken = null;
            user.ResetTokenExpires = null;
            await _dbContext.SaveChangesAsync();
            return Ok(new {
                status = true,
                message = "Password successly reset",
                data = user
                });
        }
        
        [HttpPost]
        [Authorize]
        [Route("change-password")]
        public async Task<IActionResult> ChangePassword(PasswordDto pass) {
            var user = HttpContext.User;
            var Email = user.FindFirst(ClaimTypes.Email)?.Value;
            var currentUser = _dbContext.Users.FirstOrDefault(x => x.Email == Email);
            //verify password hash
            var PasswordHasher = new PasswordHasher<string>();
            var verifPass = PasswordHasher.VerifyHashedPassword(Email, currentUser.Password, pass.OldPassword);
            if(verifPass == PasswordVerificationResult.Failed) {
                return BadRequest(new {
                status = false,
                message = "incorrect old password",
                data = new {}
                });
            }
            //hash password
            var hash = PasswordHasher.HashPassword(currentUser.Email, pass.VerifyPassword);
            currentUser.Password = hash;
            await _dbContext.SaveChangesAsync();
            return Ok(new {
                status = true,
                message = "Password Has been changed",
                data = user
                });
        }
    }
}

File: app/Http/Controllers/AuthController.php
<?php

namespace App\Http\Controllers;

use App\Helpers\HttpClient;
use GuzzleHttp\Psr7\Message;
use Illuminate\Http\Request;

class AuthController extends Controller
{
    //
    public function login(Request $request)
    {
        $payload = [
            'email' => $request->email,
            'password' => $request->password
        ];

        $auth = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/"."login",
            $payload,
        );

        if (!session()->isStarted()) {
            session()->start();
        }

        if ($auth['status'] == false) {
            return redirect('/login')->with('error', $auth['message']);
        }else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/login')->with('error', $message);
        }

        $token = $auth['data']['auth']['token'];
        $idUser = $auth['data']['user']['id'];

        session()->put("token", "$token");
        session()->put("idUser", $idUser);

        return redirect('/')->with('success', $auth['message']);
    }

    public function logout()
    {
        if (!session()->isStarted()) {
            session()->start();
        }
        session()->flush();
        return redirect('/login')->with('success',"terima kasih telah berkunjung, silahkan kembali lagi");
    }
    public function register(Request $request)
    {
        $payload = [
            'email' => $request->email,
            'password' => $request->password
        ];

        $register = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/"."register",
            $payload,
        );

        if ($register['status'] == false) {
            return redirect('/register')->with('error', $register['message']);
        } else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/register')->with('error', $message);
        }

        return redirect('/verify-email')->with('success', $register['message']);
    }

    public function verify(Request $request)
{
        $token = $request->input('token');
        
        $verify = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/verify?token=".$token,
        );

        if ($verify['status'] == false) {
            return redirect('/')->with('error', $verify['message']);
        } else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/')->with('error', $message);
        }

        return redirect('/login')->with('success', $verify['message']);
    }

    public function forgotPassword(Request $request)
    {
        $email = $request->input('email');
        $forgot = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/"."forgot-password?email=".$email,
        
        );

        if ($forgot['status'] == false) {
            return redirect('/forgot-password')->with('error', $forgot['message']);
        } else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/forgot-password')->with('error', $message);
        }

        return redirect('/reset-password')->with('success', $forgot['message']);
    }
    public function resetPassword(Request $request)
    {
        $payload = [
            'token' => $request->token,
            'password' => $request->password,
            'verifyPassword' => $request->verifyPassword,
        ];

        $reset = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/"."reset-password",
            $payload,
        );
        if ($reset['status'] == false) {
            return redirect('/reset-password')->with('error', $reset['message']);
        } else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/reset-password')->with('error', $message);
        }
        return redirect('/login')->with('success', $reset['message']);
    }
    public function changePassword(Request $request)
    {
        $payload = [
            'oldPassword' => $request->oldPassword,
            'passwordNew' => $request->passwordNew,
            'verifyPassword' => $request->verifyPassword,
        ];
        $change = HttpClient::fetch(
            "POST",
            HttpClient::apiUrl()."Auth/change-password",
            $payload,
        );
        
        if ($change['status'] == false) {
            return redirect('/change-password')->with('error', $change['message']);
        } else if(isset($reset['errors'])) {
            $errors = $reset['errors'];
            $message = array_values($errors)[0][0];
            return redirect('/change-password')->with('error', $message);
        }

        return redirect('/')->with('success', $change['message']);
    }
}
